kotlinx-coroutines-core/common/src/channels/BufferedChannel.kt

```diff
@@ -0,0 +1,3087 @@
+            is BufferedChannel<*>.BufferedChannelIterator -> tryResumeHasNextOnClosedChannel()
+            is SelectInstance<*> -> trySelect(this@BufferedChannel, CHANNEL_CLOSED)
+            else -> error("Unexpected waiter: $this")
+        }
+    }
+
+    @ExperimentalCoroutinesApi
+    override val isClosedForSend: Boolean
+        get() = isClosedForSendImpl
+
+    private val isClosedForSendImpl: Boolean
+        get() = sendersAndCloseStatus.value.isClosedForSend0
+
+    private val Long.isClosedForSend0 get() =
+        isClosed(this, sendersCur = this.sendersCounter, isClosedForReceive = false)
+
+    @ExperimentalCoroutinesApi
+    override val isClosedForReceive: Boolean
+        get() = isClosedForReceiveImpl
+
+    private val isClosedForReceiveImpl: Boolean
+        get() = sendersAndCloseStatus.value.isClosedForReceive0
+
+    private val Long.isClosedForReceive0 get() =
+        isClosed(this, sendersCur = this.sendersCounter, isClosedForReceive = true)
+
+    private fun isClosed(
+        sendersAndCloseStatusCur: Long,
+        sendersCur: Long,
+        isClosedForReceive: Boolean
+    ) = when (sendersAndCloseStatusCur.sendersCloseStatus) {
+        // This channel is active and has not been closed.
+        CLOSE_STATUS_ACTIVE -> false
+        // The cancellation procedure has been started but
+        // not linearized yet, so this channel should be
+        // considered as active.
+        CLOSE_STATUS_CANCELLATION_STARTED -> false
+        // This channel has been successfully closed.
+        // Help to complete the closing procedure to
+        // guarantee linearizability, and return `true`
+        // for senders or the flag whether there still
+        // exist elements to retrieve for receivers.
+        CLOSE_STATUS_CLOSED -> {
+            completeClose(sendersCur)
+            // When `isClosedForReceive` is `false`, always return `true`.
+            // Otherwise, it is possible that the channel is closed but
+            // still has elements to retrieve.
+            if (isClosedForReceive) !hasElements() else true
+        }
+        // This channel has been successfully cancelled.
+        // Help to complete the cancellation procedure to
+        // guarantee linearizability and return `true`.
+        CLOSE_STATUS_CANCELLED -> {
+            completeCancel(sendersCur)
+            true
+        }
+        else -> error("unexpected close status: ${sendersAndCloseStatusCur.sendersCloseStatus}")
+    }
+
+    @ExperimentalCoroutinesApi
+    override val isEmpty: Boolean get() {
+        // This function should return `false` if
+        // this channel is closed for `receive`.
+        if (isClosedForReceiveImpl) return false
+        // Does this channel has elements to retrieve?
+        if (hasElements()) return false
+        // This channel does not have elements to retrieve;
+        // Check that it is still not closed for `receive`.
+        return !isClosedForReceiveImpl
```

Comment:
We need to re-check that the channel is still not closed to obtain a correct snapshot.

Reply:
Feels like I'm missing something.

The caller entered this property, checked that `isClosedForReceiveImpl` is `false`, checked `hasElements`, it returned `false`.

Doesn't it imply that the call is already linearized and there exists a linear history where this property returned `true` even without snapshot rechecking?

Reply:
According to the `isEmpty` specification, it should return `false` if the channel is closed for receiving. 

In this implementation, we first check whether the channel is closed for receiving, followed by a check of whether it contains elements. If the channel does not have elements but is already closed, the operation should return `false` -- we need an additional `isClosedForReceive` check for that.

```
val c = BufferedChannel(capacity = 1)
c.send(1)
-----------------------
c.isEmpty() | c.cancel()
```
In the situation above, `isEmpty()` should return `false` in any case. Without the last check, it may return `true`.

Reply:
Thanks for a detailed explanation!

>In the situation above, isEmpty() should return false in any case

Should it?
It seems like linear history `send(1) -> cancel() -> isEmpty() true` is acceptable, isn't it?

Reply:
`isEmpty` should return `false` if the channel is closed for receiving.

Reply:
Oh, I see, thanks